src/player/player.c

```diff
@@ -263,6 +272,53 @@ void playerSetGrabbing(struct Player* player, struct CollisionObject* grabbing)
     } else if (grabbing != player->grabConstraint.object) {
         pointConstraintInit(&player->grabConstraint, grabbing, 8.0f, 5.0f, 1.0f);
     }
+    
+    playerInitGrabRotationBase(player);
+}
+
+void playerInitGrabRotationBase(struct Player* player) {
+    if (!player->grabConstraint.object) {
+        return;
+    }
+    struct Quaternion forwardRotation = player->lookTransform.rotation;
+    struct Vector3 forward, tmpVec, up;
+    playerGetMoveBasis(&forwardRotation, &forward, &tmpVec);
+    vector3Negate(&forward, &tmpVec);
+    quatLook(&tmpVec, &gUp, &forwardRotation);
+    playerPortalGrabTransform(player, NULL, &forwardRotation);
+    quatMultVector(&forwardRotation, &gForward, &forward);
+    quatMultVector(&forwardRotation, &gUp, &up);
+    
+    struct Quaternion objectRotation = player->grabConstraint.object->body->transform.rotation;
+    
+    // snap target rotation to nearest cube normals
+    struct Vector3 surfaceNormal, closestNormalTowards, closestNormalUp;
+    float closestNormalTowardsDot = 1.0f, closestNormalUpDot = -1.0f; // has to be negative on success
+    for (int i = 0; i < 6; ++i) {
+        quatMultVector(&objectRotation, &gCubeNormals[i], &surfaceNormal);
+        
+        float dot = vector3Dot(&surfaceNormal, &forward);
+        if (dot < closestNormalTowardsDot) {
+            closestNormalTowardsDot = dot;
+            closestNormalTowards = surfaceNormal;
+        }
+        dot = vector3Dot(&surfaceNormal, &up);
+        if (dot > closestNormalUpDot) {
+            closestNormalUpDot = dot;
+            closestNormalUp = surfaceNormal;
+        }
+    }
+    quatConjugate(&objectRotation, &objectRotation);
+    quatMultVector(&objectRotation, &closestNormalTowards, &closestNormalTowards);
+    quatMultVector(&objectRotation, &closestNormalUp, &closestNormalUp);
```

Comment:
A simplification to keep in mind for the next refactor: instead of storing normal vectors in `closestNormalTowards` and `closestNormalUp`, you could store the normal index and then undoing the rotation wouldn't be necessary. No need to change this right now since it will be refactored anyway.

Reply:
Good idea, this didn't cross my mind before due to the way I had the normal iteration set up. Since I already did the other changes anyway, I adjusted this too.